tron/serialize/runstate/DynamoDBStateStore.py

```diff
@@ -0,0 +1,132 @@
+import math
+import pickle
+from threading import Thread
+
+import boto3
+
+from tron.serialize.runstate.shelvestore import ShelveKey
+from tron.serialize.runstate.shelvestore import ShelveStateStore
+
+OBJECT_SIZE = 400000
+
+
+class DynamoDBStateStore(object):
+    def __init__(self, name, dynamodb_region) -> None:
+        self.dynamodb = boto3.resource('dynamodb', region_name=dynamodb_region)
+        self.client = boto3.client('dynamodb', region_name=dynamodb_region)
+        self.name = name
+        self.shelve = ShelveStateStore(name)
+        self.table = self.dynamodb.Table(name.replace('/', '-'))
+
+    def build_key(self, type, iden) -> ShelveKey:
+        """
+        It builds a unique partition key. The key could be objects with __str__ method.
+        """
+        return ShelveKey(type, iden)
+
+    def restore(self, keys: [ShelveKey]) -> dict:
+        """
+        Fetch all under the same parition key(keys).
+        ret: <dict of key to states>
+        """
+        def join(d1, d2):
+            for k, v in d1.items():
+                d2[k] = v
+
+        def shelve_restore(keys, res):
+            join(self.shelve.restore(keys), res)
+
+        def dynamodb_restore(keys, res):
+            items = zip(
+                keys,
+                (self[key] for key in keys),
+            )
+            join(dict(filter(lambda x: x[1], items)), res)
+
+        res1, res2 = {}, {}
+        thread1 = Thread(target = dynamodb_restore, args=(keys, res1,))
+        thread2 = Thread(target = shelve_restore, args=(keys, res2,))
+        thread1.start()
+        thread2.start()
+        thread1.join()
+        thread2.join()
+
+        # merge
+        for i, j in res1.items():
+            if i not in res2.keys():
+                res2[i] = j
+        return res2
+
+    def __getitem__(self, key: ShelveKey) -> object:
+        """
+        It returns an object which is deserialized from binary
+        """
+        val = bytearray()
+        for index in range(self._get_num_of_partitions(key)):
```

Comment:
we probably need a single query with consistent read here instead of bunch of get_item calls in a loop